### ISSUE TITLE
On-the-fly-tts keeps playing after pressing grading buttons

### ISSUE THREAD

**Issue by yaoberh** [AwesomeTTS/awesometts-anki-addon#145]
Any deck, any voice. Keeps playing until it's finished even after you press good/bad/etc and space to show the back side of the next card (audio overrides here I suppose).

**Comment by lucwastiaux** [comment]
I guess this is one of the new behaviors as a result of the integration with Anki's TTS system. If I understand correctly, you'd like it to stop playing the previous audio when you move to the new card ?

**Comment by yaoberh** [comment]
Yes, it doesn't stop playing when you grade the card and it's pretty distracting when you try to read the next one and the previous audio is still playing. I thought it's a bug, at least in my opinion.

**Comment by lucwastiaux** [comment]
OK, I started a thread on the anki forum here: https://forums.ankiweb.net/t/anki-3rd-party-tts-service-how-to-stop-playing-audio/4799 , I probably need help from Damien on this.

**Comment by yaoberh** [comment]
This error/bug is still happening. Have any new discoveries as to why been made?

**Comment by lucwastiaux** [comment]
Could you send me your deck and give me an example of a card where this happens ? From the Anki main screen, click File → Export, click Export... . Then locate apkg file you just created, and upload it to https://www.dropbox.com/request/X9L4mhPqEcfvKFuSm2nK

**Comment by yaoberh** [comment]
Done! I uploaded a subdeck of my main japanese profile. This happens on all my profiles and it seems to be only with on-the-fly tts. Not with generated mp3's. It's a very strange behaviour because when you see a card and the audio hasn't started playing but you already grade the card, then the audio starts playing but it's the one of the card you just graded. It overrides with the audio of the 3rd one because of some threading functionality I suppose.

**Comment by yaoberh** [comment]
Hey there, are there any news on this error? It hasn't gone away since I reported it and I've been upgrading anki and nothing has changed. Is there anything I could try? Thanks!

**Comment by shubniguroth** [comment]
So one way I've personally gotten around this is by using Automator on MacOS to pre-generate the on-the-fly files each day.

I made an Automator script that presses space bar in Anki every 2 and then every 4 seconds. And then it repeats this 600-700 times. Then it presses Cmd+Z 600-700 times. In this way, I force AwesomeTTS to pre-cache the tag files. I run it every morning while I'm doing other things, and then when I get to my Anki, it's all already there. 

 I use Google's Wavenet TTS, don't know if that matters to the timing. I think it's very specific to the length of your front and back text. 2 and 4 seconds seems to cover my edge cases pretty well.

This overriding of audio described in the OP only seems to happen when it's a newly encountered non-cached corpus, so this workaround means I no longer have audio over-playing the next card. It's a kind of hybrid between the batch generation and on-the-fly context generation.

Just my 2 cents. A temporary fix. Good luck y'all.

**Comment by yaoberh** [comment]
Hey, that's a lot of effort! I'm on linux here though, and anyhow this wouldn't suit me because I have A LOT of different voices of many different services and they all differ in text length...So I hope this gets fixed some day :D

**Comment by lucwastiaux** [comment]
so if I understand correctly, this is due to lag when generating the audio which is not cached ?

**Comment by yaoberh** [comment]
Yep, seems to be cache. You're right. Happens mainly when one reviews a card and graduates it before the audio has gotten the time to play and it just keeps on playing even after the next card shows up. After this it follows with the proper audio of the card being shown.

**Comment by lucwastiaux** [comment]
understood. and what service are you using to generate the audio ? is it a service which is particularly slow ?

**Comment by yaoberh** [comment]
Nope, no service in particular. It appears to happen with all services. I just use Naver Papago at the moment for Chinese.
